src/charm.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm responsible for distributing certificates through relationship.
+
+Certificates are provided by the operator trough Juju configs.
+"""
+
+import logging
+from typing import List
+
+from charms.tls_certificates_interface.v0.tls_certificates import (
+    Cert,
+    CertificateRequestEvent,
+    TLSCertificatesProvides,
+)
+from ops.charm import CharmBase, ConfigChangedEvent
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus
+
+logger = logging.getLogger(__name__)
+
+
+class TLSCertificatesOperatorCharm(CharmBase):
+    """Main class to handle Juju events."""
+
+    def __init__(self, *args):
+        """Observes config change and certificate request events."""
+        super().__init__(*args)
+        self.tls_certificates = TLSCertificatesProvides(self, "certificates")
+        self.framework.observe(
+            self.tls_certificates.on.certificate_request, self._on_certificate_request
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Triggered once when the charm is installed.
+
+        Args:
+            event (ConfigChangedEvent): Juju event.
+
+        Returns:
+            None
+        """
+        missing_config_options = self.get_missing_configuration_options()
+        if missing_config_options:
+            self.unit.status = BlockedStatus(
+                f"Configuration options missing: {missing_config_options}"
+            )
+        else:
+            self.unit.status = ActiveStatus()
+
+    def _on_certificate_request(self, event: CertificateRequestEvent) -> None:
+        """Triggered everytime there's a certificate request.
+
+        Args:
+            event (CertificateRequestEvent): Event for certificate requests.
+
+        Returns:
+            None
+        """
+        if self._certificates_are_set:
+            self.tls_certificates.set_relation_certificate(
+                certificate=Cert(
+                    cert=self.model.config["certificate"],
+                    key=self.model.config["private-key"],
+                    ca=self.model.config["ca-certificate"],
+                    common_name=event.common_name,
```

Comment:
This seems a bit simplistic, as the certificate that we will return might not match the `common_name` that was requested. Maybe we could open another ticket to improve this by checking if the certificate that we have matches the request before sending it.

Reply:
True! Right now this charm doesn't know about the common name though, to do this we would need to add logic to parse the certificate provided by the user.